### PR TITLE
Support latest version

### LINKED ISSUE #266: Support latest version tag
#### Is your feature request related to a problem?

There is no way to specify the latest version, such as `ubuntu-latest`. 

#### Describe the feature you'd like

Being able to retrieve the latest version of Elixir and Erlang OTP:
```
elixir:  latest
otp: latest
```

#### Describe alternatives you've considered

Currently using main and maint, but those branches are in constant development, and I want to be able to test on the latest release.

### PR DESCRIPTION
**Pull request by Schultzer**
# Description

This PR adds support for the latest version; I'm not super familiar with the codebase, so it's not the prettiest, and the tests are doomed to break when there is a new release.

@paulo-ferraz-oliveira Please let me know if this is the right direction.

Closes #266 

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/erlef/setup-beam/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
If we want to support this we need to do it for all dependencies we maintain (includes Gleam, rebar3, etc.), for feature parity.

**Comment by paulo-ferraz-oliveira** [comment]
> the tests are doomed to break when there is a new release.

We're usually up-to-date with newer releases, and update the action to test those, where appropriate.

File: src/setup-beam.js [hunk 1]
```diff
@@ -250,6 +250,12 @@ async function getRebar3Version(r3Spec) {
   return rebar3Version
 }
 
+async function getLatestOTPVersions() {
+  const response = await fetch("https://api.github.com/repos/erlang/otp/releases/latest")
+  const json = await response.json();
+  return json.tag_name
+}
```
Comment (paulo-ferraz-oliveira) [inline src/setup-beam.js:257]: This would look much better as a single function that took "org/repo" as input.

**Comment by paulo-ferraz-oliveira** [comment]
> I have read and understood the [contributing guidelines](https://github.com/erlef/setup-beam/blob/main/CONTRIBUTING.md)

This states "Execute `npm run build-dist` and fix any issues arising from that". I can't understand what failed that it didn't generate any artifacts that you'd need to push, for which CI complained later, for example [here](https://github.com/erlef/setup-beam/actions/runs/8953653372/job/24605977124?pr=267).

**Comment by Schultzer** [comment]
Thanks, @paulo-ferraz-oliveira I've cleaned this up.

**Comment by ericmj** [comment]
Should it be the latest stable version or include pre-releases?

You need to fetch the latest version from the same source we get the other versions, otherwise the task will fail when there is a new release as it's tagged before the builds are available (at least for elixir and erlang).

**Comment by Schultzer** [comment]
> Should it be the latest stable version or include pre-releases?
> 
> You need to fetch the latest version from the same source we get the other versions, otherwise the task will fail when there is a new release as it's tagged before the builds are available (at least for elixir and erlang).

Do you have an example of this, and how long would it take to create a release before the builds are available? 


https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-the-latest-release 

> The latest release is the most recent non-prerelease, non-draft release, sorted by the created_at attribute. The created_at attribute is the date of the commit used for the release, and not the date when the release was drafted or published.

The latest release does not include pre-releases.

**Comment by ericmj** [comment]
> Do you have an example of this

You can check the `builds.txt` files we fetch to get the list of versions.

> and how long would it take to create a release before the builds are available?

It depends on a lot of factors but normally 10-30min. Sometimes a build fails which requires manual fixing which can take hours to days.

**Comment by paulo-ferraz-oliveira** [comment]
Let's say `erlang/otp` releases 24.2.5, but 24.2.4 is the one set to "latest". Shall a rule in `setup-beam` override that? Or where will the semantics reside?

**Comment by Schultzer** [comment]
> You can check the builds.txt files we fetch to get the list of versions.

I've refactored to get the latest semver version from the build.text, which should solve the issue with builds not being available and the test suite passing locally for me.

**Comment by paulo-ferraz-oliveira** [comment]
Let's say `erlang/otp` releases 24.2.5, but 24.2.4 is the one set to "latest". Shall a rule in `setup-beam` override that? Or where will the semantics reside?

**Comment by Schultzer** [comment]
> Let's say `erlang/otp` releases 24.2.5, but 24.2.4 is the one set to "latest". Shall a rule in `setup-beam` override that? Or where will the semantics reside?

I realized I had a few bugs, so now I use Semver to resolve the latest Elixir and OTP builds. Furthermore, a bug in Semver prevented prereleases from being parsed, so I've updated the package.

Since we know if the tag has a top in its name, we can leverage Semver to parse that with its right-hand rule.

I tested this locally by changing the platform clauses from Linux to Darwin and Win32 to Darwin. I'm confident about the changes I committed.

**Comment by ericmj** [comment]
How does the semver library handle OTP versions that does not follow the semver specification, such as `OTP-25.3.2.12`?

**Comment by Schultzer** [comment]
I got lost in this codebase and the review comments; I've opened #269, which is way simpler than I tried to do here.

**Pull request closed without merging**